external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaUtils.scala

```diff
@@ -144,4 +150,249 @@ object KafkaUtils {
+   * @param fromOffsets per-topic/partition Kafka offsets defining the (inclusive)
+   *  starting point of the stream
+   * @param maxRetries maximum number of times in a row to retry getting leaders' offsets
+   */
+  @Experimental
+  def createNewStream[
+    K: ClassTag,
+    V: ClassTag,
+    U <: Decoder[_]: ClassTag,
+    T <: Decoder[_]: ClassTag,
+    R: ClassTag] (
+      ssc: StreamingContext,
+      kafkaParams: Map[String, String],
+      fromOffsets: Map[TopicAndPartition, Long],
+      messageHandler: MessageAndMetadata[K, V] => R,
+      maxRetries: Int
+  ): InputDStream[R] = {
+    new DeterministicKafkaInputDStream[K, V, U, T, R](
+      ssc, kafkaParams, fromOffsets, messageHandler, maxRetries)
+  }
+
+  /**
+   * This stream can guarantee that each message from Kafka is included in transformations
+   * (as opposed to output actions) exactly once, even in most failure situations.
+   *
+   * Points to note:
+   *
+   * Failure Recovery - You must checkpoint this stream.
+   * Kafka must have sufficient log retention to obtain messages after failure.
+   *
+   * Getting offsets from the stream - see programming guide
+   *
+.  * Zookeeper - This does not use Zookeeper to store offsets.  For interop with Kafka monitors
+   * that depend on Zookeeper, you must store offsets in ZK yourself.
+   *
+   * End-to-end semantics - This does not guarantee that any output operation will push each record
+   * exactly once. To ensure end-to-end exactly-once semantics (that is, receiving exactly once and
+   * outputting exactly once), you have to ensure that the output operation is idempotent.
+   *
+   * @param ssc StreamingContext object
+   * @param kafkaParams Kafka <a href="http://kafka.apache.org/documentation.html#configuration">
+   * configuration parameters</a>.
+   *   Requires "metadata.broker.list" or "bootstrap.servers" to be set with Kafka broker(s),
+   *   NOT zookeeper servers, specified in host1:port1,host2:port2 form.
+   *   If starting without a checkpoint, "auto.offset.reset" may be set to "largest" or "smallest"
+   *   to determine where the stream starts (defaults to "largest")
+   * @param topics names of the topics to consume
+   */
+  @Experimental
+  def createNewStream[
+    K: ClassTag,
+    V: ClassTag,
+    U <: Decoder[_]: ClassTag,
+    T <: Decoder[_]: ClassTag] (
+      ssc: StreamingContext,
+      kafkaParams: Map[String, String],
+      topics: Set[String]
+  ): InputDStream[(K, V)] = {
+    val messageHandler = (mmd: MessageAndMetadata[K, V]) => (mmd.key, mmd.message)
+    val kc = new KafkaCluster(kafkaParams)
+    val reset = kafkaParams.get("auto.offset.reset").map(_.toLowerCase)
+
+    (for {
+      topicPartitions <- kc.getPartitions(topics).right
+      leaderOffsets <- (if (reset == Some("smallest")) {
+        kc.getEarliestLeaderOffsets(topicPartitions)
+      } else {
+        kc.getLatestLeaderOffsets(topicPartitions)
+      }).right
+    } yield {
```

Comment:
The method you're commenting on doesn't take an offset, and the doc already
says it doesn't store offsets in zookeeper, so there is no other valid
offset to be used.  It will just use whatever is specified by
auto.offset.reset
On Feb 1, 2015 8:09 PM, "Saisai Shao" <notifications@github.com> wrote:

> In
> external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaUtils.scala
> <https://github.com/apache/spark/pull/3798#discussion_r23904486>:
>
> > +      ssc: StreamingContext,
> > +      kafkaParams: Map[String, String],
> > +      topics: Set[String]
> > +  ): InputDStream[(K, V)] = {
> > +    val messageHandler = (mmd: MessageAndMetadata[K, V]) => (mmd.key, mmd.message)
> > +    val kc = new KafkaCluster(kafkaParams)
> > +    val reset = kafkaParams.get("auto.offset.reset").map(_.toLowerCase)
> > +
> > +    (for {
> > +      topicPartitions <- kc.getPartitions(topics).right
> > +      leaderOffsets <- (if (reset == Some("smallest")) {
> > +        kc.getEarliestLeaderOffsets(topicPartitions)
> > +      } else {
> > +        kc.getLatestLeaderOffsets(topicPartitions)
> > +      }).right
> > +    } yield {
>
> Is this "auto.offset.reset" just a hint for out-of-range re-position of
> offset, or just a immediate action to seek the offset to beginning or end?
> As I know this configuration is just a hint, will not be effective when the
> offset is valid, I'm not sure here your behavior, is the same as Kafka?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/3798/files#r23904486>.
>